File: 백준/Gold/17298. 오큰수/오큰수.cc
#include <stdio.h>
#include <iostream>
#include <string>
#include <stack>
#include <utility>
int main(){
    std::cin.tie(0);
    std::ios::sync_with_stdio(0);
    std::stack <std::pair<int,int>> stack{};
    int n{};
    long num{};
    std::cin >> n;
    int answer[n]{};
    for(int i=0; i<n; i++){
        answer[i]=-1;
    }
    for(int i=0; i<n; i++){
        std::cin >> num;
        while(!stack.empty() && stack.top().first<num){
            answer[stack.top().second]=num;
            stack.pop();

        }
        stack.push(std::make_pair(num,i));
    }
    for(int i: answer){
        std::cout<< i<<" ";
    }
}

File: 백준/Gold/17298. 오큰수/오큰수.py
import sys
n=int(sys.stdin.readline())
answer=[-1 for i in range(n)]
arr=[]
num=list(map(int,sys.stdin.readline().split()))
for i in range(n):
    while(arr and arr[-1][0]<num[i]):
        answer[arr[-1][1]]=num[i]
        arr.pop()
    arr.append([num[i],i])
print(*answer)